<a href="https://colab.research.google.com/github/IyadSultan/educational/blob/main/Python_for_Pediatric_Oncology.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Python for Pediatric Oncology: Introductory Coding Workshop

# Variables
Variables are like labeled containers for information. In medicine, you might put a patient’s age in one field, weight in another – in Python, you’d store these values in variables. A variable has a name and holds some value (number, text, etc.). You can then use that name to retrieve or change the value. For example, let's create a variable for a patient's age and another for their name:

In [ ]:
# Storing patient info in variables
patient_name = "Alice"
patient_age = 7
print("Patient name:", patient_name)
print("Patient age:", patient_age)


## Data types
Python has several basic data types:
Integers (int) for whole numbers (e.g., 7 for age).
Floats for decimals (e.g., 7.5 for weight in kg perhaps).
Strings (str) for text (e.g., "Alice" for a name or "Stage I" for a disease stage).
Booleans (bool) for True/False values (e.g., True for has fever, False for no fever).

### Lists

In [ ]:
patient_names = ["Alice", "Bob", "Charlie"]
temperatures = [36.6, 38.0, 39.4]  # in °C
print(patient_names)
print("First patient:", patient_names[0])


This creates a list of three names and a list of three temperature readings. patient_names[0] accesses the first element ("Alice") since indexing starts at 0 in Python. Lists are super useful to hold multiple related items (like records in a cohort).

Analogy: Think of a variable as a labeled jar where you can put a piece of data. You might have a jar labeled "Age" containing the number 7. You can open it and replace it with 8 later. A list is like a pill organizer with multiple slots – each slot can hold a value, and you can access each by its position.

### Exercise 1
Generate a list of 10 random numbers (between 1 and 6) and print the first and last number

Make a variable hospital_name for your hospital and print it.

In [ ]:
# prompt: Make a variable hospital_name and give it a value of "KHCC"

import random

# Generate a list of 10 random numbers between 1 and 6
random_numbers = [random.randint(1, 6) for _ in range(10)]

# Print the first and last numbers
print("First number:", random_numbers[0])
print("Last number:", random_numbers[-1])

# Make a variable hospital_name and give it a value of "KHCC"
hospital_name = "KHCC"
print("Hospital Name:", hospital_name)


Create a list ward_numbers with some ward/floor numbers and try printing the second item (index 1).

In [3]:
# prompt: Create a list ward_numbers with some ward/floor numbers and try printing the second item (index 1).

ward_numbers = [101, 202, 303, 404]
print("Second ward/floor number:", ward_numbers[1])


Second ward/floor number: 202


Change one element in your list (e.g., update a ward number) and print the list again to see the change.

In [4]:
# prompt: Change one element in your list by changing the first element to 102 and print the list again to see the change.

ward_numbers[0] = 102
ward_numbers


[102, 202, 303, 404]

# Loops and Conditional Logic

## Loops
In programming, loops let us repeat actions easily, and conditional statements (if/else) let us make choices based on data. These are like the bread and butter of automating tasks. For Loops: A for-loop allows you to iterate over a sequence (like each item in a list) and do something with each item. Imagine doing morning rounds for each patient in a list – that's a loop in action (repeating a procedure for each patient). Let's say we have a list of patient temperatures and we want to check each for fever:

```
# This is formatted as code
```



In [5]:
temperatures = [36.6, 38.0, 39.4]  # three patients' temperatures
for temp in temperatures:
    print("Current temp:", temp)


Current temp: 36.6
Current temp: 38.0
Current temp: 39.4


## If conditions
 Now, just printing isn't that useful – let's add a decision. We can use an if statement inside the loop to flag which temperatures indicate fever. In medicine, we might say >37.5°C is a fever. We can code that logic:


In [6]:
for temp in temperatures:
    if temp > 37.5:
        print(temp, "-> High fever! 🔥")
    else:
        print(temp, "-> Normal")


36.6 -> Normal
38.0 -> High fever! 🔥
39.4 -> High fever! 🔥


We introduced an if/else:

- If the condition temp > 37.5 is true, we execute the first block (print "High fever! 🔥").

- Otherwise (else), we execute the second block (print "Normal").

The 🔥 emoji is just for fun to highlight fevers.
(Yes, you can include emojis in Python strings!) This small snippet combined a loop and a conditional check it went through each temperature and made a decision for each. You could imagine extending this: e.g., if fever, maybe alert the user or increment a counter of how many fevers were found, etc.

*Analogy*: Loops are like doing the same test on every sample in a batch, or like a nurse checking each patient in a ward one by one. If/else statements are like triaging: if a patient is critical, send to ICU, else send them to a normal ward.

### Ecercise 2

Add a new temperature to the list (e.g., 37.0) and re-run the loop. Does it correctly label it as normal?

Change the fever threshold to 38.0 in the code (replace 37.5 with 38.0) and run again. Now 38.0°C would be considered normal (because it’s not greater than 38.0). This shows how a simple code change can adjust your logic.

Extra: Create a list of blood pressure readings and write a loop to print which ones are above a certain threshold (you decide the threshold).

In [7]:
# prompt: Create a list of blood pressure readings (100/50, 60/40, 110/60) and write a loop to print which ones are above a certain threshold where mean arterial pressure < 50.

blood_pressure_readings = ["100/50", "60/40", "110/60"]
threshold = 50  # Mean Arterial Pressure threshold

for reading in blood_pressure_readings:
    systolic, diastolic = map(int, reading.split("/"))
    map_value = (1/3) * systolic + (2/3) * diastolic
    if map_value < threshold:
        print(f"Blood pressure reading {reading} is below the threshold (MAP: {map_value})")


Blood pressure reading 60/40 is below the threshold (MAP: 46.666666666666664)


# Functions (Reusable Code Blocks)

Functions are a way to package a set of instructions so you can reuse them with different inputs. Think of a function like a recipe or a medical protocol: you define it once, and then you can use it whenever needed, without rewriting all the steps. In Python, we define a function using the def keyword, give it a name, parameters (inputs) in parentheses, and a block of code. For example, let's define a simple function to categorize a Hodgkin lymphoma stage into "Early-stage" vs "Advanced-stage" disease:

In [9]:
def classify_stage(stage):
    if stage <= 2:
        return "Early-stage"
    else:
        return "Advanced-stage"


This function classify_stage takes one input (stage). Inside, it uses an if/else to decide the return value. If stage is 1 or 2, it returns "Early-stage"; if 3 or 4, it returns "Advanced-stage". Notice the indentation – Python uses indentation (spaces) to define blocks of code under the function and under the if/else. Colab will help by auto-indenting after you type a : and press Enter.

In [10]:
print(classify_stage(1))  # Stage I -> expected "Early-stage"
print(classify_stage(4))  # Stage IV -> expected "Advanced-stage"


Early-stage
Advanced-stage


We can call classify_stage for any stage number now, and it will consistently apply the rule we wrote. Functions can have multiple parameters as well. For instance, you might have a function to calculate Body Mass Index:

In [11]:
def calculate_bmi(weight_kg, height_m):
    bmi = weight_kg / (height_m ** 2)
    return bmi

print(calculate_bmi(50, 1.6))  # 50 kg, 1.6 m height


19.531249999999996


This would output a BMI value (around 19.5 for those numbers). The details of BMI aren't central here, but it shows how a function can take inputs and produce an output (using return). The return statement hands back the result to wherever the function was called.

*Analogy*: A function is like a diagnostic test – you provide a sample (input), and it gives you a result (output) after processing. Once a test is developed, any patient sample can go through it to get a result, without redesigning the test each time. By organizing code into functions, you make it more modular and readable. If you have a complex calculation you'll do often, define it as a function and then your main code becomes much cleaner (“just call the function”).

### Exercise 3

Define a function greet_doctor(name) that takes a name and returns a greeting like "Hello Dr. <name>, welcome!"

Test your function by calling it with a few different names.

Modify the classify_stage function to handle an edge case: if stage is 0 or not 1-4, have it return "Unknown stage" (hint: use an if before the others to check if stage not in [1,2,3,4]).

# Using Libraries (Importing Modules)

One of Python’s superpowers is its rich ecosystem of libraries (also called modules or packages). Libraries are like add-on toolkits that provide extra functions and capabilities. For example, there are libraries for scientific computing, for making plots, for machine learning, and much more. We'll use some libraries later for data analysis (like pandas and matplotlib).

To use a library, you first import it. Python has many built-in libraries (like math, datetime, etc.), and thousands of external ones you can install. In Colab, many common libraries (numpy, pandas, matplotlib, etc.) are pre-installed.

Let's see a quick example with the built-in math library:

In [ ]:
import math
value = math.sqrt(16)
print("Square root of 16 is:", value)

In [16]:
from math import sqrt
value = sqrt(16)
print("Square root of 16 is:", value)


Square root of 16 is: 4.0
Square root of 16 is: 4.0


This should output Square root of 16 is: 4.0. We did import math (bringing in the math module), and then used math.sqrt() function to compute a square root. The math library has many math functions (try math.log, math.sin, etc.).

We’ll soon use pandas (for data handling) and matplotlib (for plotting). When we do, we’ll write import pandas as pd and import matplotlib.pyplot as plt – the as pd and as plt are just aliases to save typing.

*Note* : You only need to import a library once at the top of your notebook or code. After that, you can use its functions.

Libraries allow you to accomplish complex tasks with just a few lines of code because others have developed and tested these tools for you. This is especially useful in medical data analysis – why write a statistical function from scratch if a trusted library can do it?

### Exercise 4

import random and then use random.choice([list]) to pick a random element from a list of your choice (e.g., random.choice of a list of patient names).

In [22]:
# prompt: import random and then use random.choice([list]) to pick a random element from a list of my patients (John, Samir, Helen)

random_patient = random.choice(["John", "Samir", "Helen"])
random_patient


'John'

import datetime and use datetime.datetime.now() to get the current date and time (just to see an example of using a slightly larger library).

In [ ]:
# prompt: import datetime and use datetime.datetime.now() to get the current date and time

import datetime

# ... (rest of your existing code)

# Get the current date and time
current_datetime = datetime.datetime.now()
print("Current date and time:", current_datetime)


just type "import this" for fun

In [23]:
import this

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!
